exist-core/src/main/java/org/exist/test/XQueryAssertions.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * eXist-db Open Source Native XML Database
+ * Copyright (C) 2001 The eXist-db Authors
+ *
+ * info@exist-db.org
+ * http://www.exist-db.org
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+package org.exist.test;
+
+import com.evolvedbinary.j8fu.Either;
+import org.exist.xquery.CompiledXQuery;
+import org.exist.xquery.ErrorCodes;
+import org.exist.xquery.XPathException;
+import org.exist.xquery.value.Sequence;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * This set of assertions are meant to help when testing XQuery compilation, execution and errors
+ * Especially useful, if the testsuite inherits from/ extends XQueryCompilationTest class
+ * @author <a href="mailto:juri@existsolutions.com">Juri Leino</a>
+ */
+public class XQueryAssertions {
+    public static void assertXQStaticError(final ErrorCodes.ErrorCode expectedCode, final int line, final int column, final String expectedMessage, final Either<XPathException, CompiledXQuery> actual) {
+        assertXQStaticError(expectedCode, line, column, actual);
+        assertXQErrorMessage(expectedMessage, actual.left().get());
+    }
+
+    public static void assertXQStaticError(final ErrorCodes.ErrorCode expectedCode, final int line, final int column, final Either<XPathException, CompiledXQuery> actual) {
+        assertTrue("Expected: " + expectedCode.getErrorQName() + ", but got result: " + actual.toString(), actual.isLeft());
+        final XPathException xpe = actual.left().get();
+        assertXQErrorCode(expectedCode, xpe);
+        assertXQErrorLine(line, xpe);
+        assertXQErrorColumn(column, xpe);
+    }
+
+    public static void assertXQDynamicError(final ErrorCodes.ErrorCode expectedCode, final int line, final int column, final String expectedMessage, final Either<XPathException, Sequence> actual) {
+        assertXQDynamicError(expectedCode, line, column, actual);
+        assertXQErrorMessage(expectedMessage, actual.left().get());
+    }
+
+    public static void assertXQDynamicError(final ErrorCodes.ErrorCode expectedCode, final int line, final int column, final Either<XPathException, Sequence> actual) {
```

Comment:
This appears to be a copy of `assertXQStaticError`. If (as it looks to me) I`assertXQStaticError` and `assertXQDynamicError` are indeed the same, I would ask that your remove the duplicate code and rename it to just `assertXQueryError`. I can't see anything that actually asserts whether it is a dynamic or static error... or did I miss something?

Reply:
Yep, the important part. One compiles the other executes the query.

Reply:
DynamicErrors have `Either<XPathException, Sequence>` as actual
StaticErrors have `Either<XPathException, CompiledXQuery>`

Reply:
Ah okay! I see...

Just wondering if you had considered:

1. In this code `actual.toString()` is eagerly (always) evaluated regardless:
```java
assertTrue("Expected: " + expectedCode.getErrorQName() + ", but got result: " + actual.toString(), actual.isLeft());
```
The problem here is that the `toString()` could raise an exception. For example where Sequence is a DocumentImpl which has been disposed, toString raises an exception.

2. `toString()` may not be implemented for all implementations of the `CompiledXQuery` and `Sequence` interfaces... so you may get a non-sensical error message in those cases.

You could benefit from a better formulation like:
```java
assertLeft(() -> "Expected: " + expectedCode.getErrorQName() + ", but got result: " + actual.toString(), actual);

public static <L,R> void assertLeft(final Supplier<String> message, final Either<L,R> either) {
  if (!either.isLeft()) {
    fail(message.get());
  }
}
```

Reply:
If `R.toString()` might not be present, how is the above helping. Calling `message.get()` would still throw in that case, no?

Reply:
Wouldn't a static failure message be a safer option. Something along the lines of
```
if (!either.isLeft()) {
  fail("Not an error")
}
```

Reply:
I would settle for 
```java
    public static <L,R> void assertIsError(final ErrorCodes.ErrorCode expected, final Either<L,R> either) {
        if (!either.isLeft()) {
            fail("Expected error: " + expected.getErrorQName() + ", but no error was thrown.");
        }
    }
```

and 

```java
    public static void assertXQDynamicError(final ErrorCodes.ErrorCode expectedCode, final int line, final int column, final Either<XPathException, Sequence> actual) {
        assertIsError(expectedCode, actual);
        final XPathException xpe = actual.left().get();
        assertXQErrorLine(line, xpe);
        assertXQErrorColumn(column, xpe);
        assertXQErrorCode(expectedCode, xpe);
    }
```

Reply:
@adamretter please re-review

Reply:
> If <R>.toString() might not be present, how is the above helping. Calling message.get() would still throw in that case, no?

It's not that it's not present, it's that it would raise an exception. The difference is, that when the tests pass, you won't get an exception on the toString. You are correct, that when the tests fail, even with the new formulation, you might get an exception on the toString, so you may have to diagnose the compound failure... but a failure is a failure so it's not so bothersome in that position.

Reply:
wouldn't that still evaluate the other assertions if it we do not explicitly fail here? That's how understood your previous comment.

Reply:
@line-o Can you elaborate, I don't understand your question.

Reply:
What happens if `assertTrue` in `assertIsError` fails? Will the following assertions run or will the test runner stop at the first failing assertion as fail would definitely do.

Reply:
Ah, I think we are talking at cross-purposes. I was saying you can get rid of `assertIsError` completely now and just use the final `assertTrue` that I posted.

Reply:
That does not answer my question and would also lead to duplicate code. I do like the current solution and would keep it.

Reply:
You wrote:

> What happens if assertTrue in assertIsError fails.

As I am saying that there would be no `assertIsError` anymore, I don't see a question that I can answer. Do you have a new question or did I miss something?

Reply:
> Will the following assertions run or will the test runner stop at the first failing assertion as fail would definitely do?

Reply:
And if not, I would like to avoid to write `if (!either.isLeft()) { fail(message) }` twice

Reply:
ping?

Reply:
OK, I tested your proposed solution and it works as expected. The other assertions are not being evaluated. This was changed in the XQueryAssertions class.